Migrate Cover Block tests to Playwright

worldomonation opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR is part of the migration effort tracked at https://github.com/WordPress/gutenberg/issues/38851.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This PR proposes a migration of the `cover.test.js` to the new Playwright-based framework.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
`npm run test:e2e:playwright -- test/e2e/specs/editor/blocks/cover.spec.js`

In the current form (pending outcome of [this question](https://github.com/WordPress/gutenberg/pull/45784#issuecomment-1523034256), it should produce the following outcome:

```
Running 6 tests using 1 worker

  ✓  1 …2 › Cover › can set overlay color using color picker on block placeholder (2.4s)
Failed to load resource: the server responded with a status of 404 (Not Found)
  ✓  2 … Cover › can set background image using image upload on block placeholder (3.3s)
Failed to load resource: the server responded with a status of 404 (Not Found)
  ✓  3 …over.spec.js:102:2 › Cover › dims background image down by 50% by default (1.8s)
Failed to load resource: the server responded with a status of 404 (Not Found)
  ✓  4 …] › editor/blocks/cover.spec.js:126:2 › Cover › can have the title edited (2.3s)
Failed to load resource: the server responded with a status of 404 (Not Found)
  ✓  5 …tor/blocks/cover.spec.js:151:2 › Cover › can be resized using drag & drop (2.4s)
Failed to load resource: the server responded with a status of 404 (Not Found)
  -  6 …er › dims the background image down by 50% when transformed from the Image block

  1 skipped
  5 passed (13.1s)
```

## Screenshots or screencast <!-- if applicable -->

worldomonation committed:
Add new Playwright E2E tests for the cover block.

worldomonation committed:
Add comment

worldomonation commented:
Hi folks - this is my first PR in the Gutenberg repo. As I understand it, I must wait for the tests to be triggered manually.

worldomonation commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,245 @@
+			.click();
+
+		// Open the block list viewer and select child paragraph within the cover block.
+		await page.locator( '[aria-label="List View"]' ).click();
```
**question**: I've chosen to use `aria-label` attributes instead of CSS selectors on several occasions in this file, building on past experience and the [Best Practices](https://playwright.dev/docs/selectors#best-practices) guideline from the Playwright maintainers. 

However, I see that Gutenberg repo tends to use CSS selectors under most scenarios. Is this an official guideline?

kevin940726 replied:
We have our own guidelines and [best practices](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md#use-accessible-selectors). The general recommendation is to use accessible selectors whenever possible. It should be really rare to resort to CSS selectors. There are several places still using CSS selectors, but they can be considered legacy code. In this case, I think a role selector will work best!

worldomonation replied:
@kevin940726 Thanks for the heads up. I think the code style was lost on me during the division meetup; I will look at updating this PR and re-request a review once done.

worldomonation commented:
Looks like my run at https://github.com/WordPress/gutenberg/actions/runs/3474015855 failed. I will need to investigate why it is failing on CI and not locally.

worldomonation committed:
Update block list selector

kevin940726 commented:
Hi! I only just saw this now 😞 . Sorry for the late reply! And thank you for your contribution!

worldomonation commented:
@kevin940726 I have a quick question regarding the `await editor.transformBlockTo( <block_name> )` method.

When I call the aforementioned method in one of the specs I am migrating, the following error is thrown at the end:

```
  1) [chromium] › editor/blocks/cover.spec.js:200:7 › Cover › dims the background image down by 50% when transformed from the Image block

    Error: page.evaluate: window.wp.galleryBlockV2Enabled is not defined

       at ../../../packages/e2e-test-utils-playwright/src/editor/transform-block-to.ts:13

      11 |  */
      12 | export async function transformBlockTo( this: Editor, name: string ) {
    > 13 | 	await this.page.evaluate(
         | 	                ^
      14 | 		( [ blockName ] ) => {
      15 | 			const clientIds = window.wp.data
      16 | 				.select( 'core/block-editor' )

        at Editor.transformBlockTo (gutenberg/packages/e2e-test-utils-playwright/src/editor/transform-block-to.ts:13:18)
        at gutenberg/test/e2e/specs/editor/blocks/cover.spec.js:221:16
```

I'm do not understand why the `transformBlockTo` is calling `galleryBlockV2Enabled`. Could you point me in the right direction?

worldomonation committed:
test/e2e/specs/editor/blocks/cover.spec.js
- update spec with accessibility-based selectors where possible.

kevin940726 commented:
@worldomonation Could you share a branch so that I can test locally?

worldomonation commented:
> @worldomonation Could you share a branch so that I can test locally?

Sure, my work is in the forked repo `worldomonation:add/playwright-cover-block-e2e`. 
[Link to the branch in my forked repo](https://github.com/worldomonation/gutenberg/tree/add/playwright-cover-block-e2e).

kevin940726 commented:
@worldomonation I just checked out your code and the test passed locally for me. Could you try again?

worldomonation commented:
> @worldomonation I just checked out your code and the test passed locally for me. Could you try again?

Thanks for trying. I had marked the step with the issue, `dims the background image down by 50% when transformed from the Image block`, as skipped. Did you unskip it?

kevin940726 commented:
@worldomonation Yep, I unskipped it.

worldomonation commented:
> @worldomonation Yep, I unskipped it.

I think I figured it out. 
I started this PR back at Denver in late 2022, and haven't looked at it until this week. I thought I had updated my local, forked repo with upstream but looks like I neglected to do that plus a `npm run build`.

It now passes for me locally too. Thanks for helping me work through the issues.

I hope to have this ready for review soon.

worldomonation committed:
test/e2e/specs/editor/blocks/cover.spec.js
- rebase, rebuild and update the selectors

worldomonation committed:
packages/e2e-tests/specs/editor/blocks/cover.test.js
- remove Puppeteer file.

worldomonation commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,242 @@
+		const coverBlock = page.getByRole( 'document', {
+			name: 'Block: Cover',
+		} );
+		const coverBlockResizeHandle = page.locator(
+			'.components-resizable-box__handle-bottom'
```
Despite preferring accessible attributes as selectors, it turns out the draggable handler on the bottom edge of the Cover block doesn't have anything that can be used to select it other than the CSS selector.

kevin940726 replied:
Yeah, it's unfortunate, and we can't add testid either. Adding a comment above explaining the situation should help though!

worldomonation replied:
Comment added in 071c31b4fe4685041cbd67bfe90e7d5f27131a0c.

worldomonation committed:
test/e2e/specs/editor/blocks/cover.spec.js
- clarify use of `span` element in the first spec.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+async function getTestImage() {
```
Could we follow what other [tests](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/specs/editor/blocks/image.spec.js) do and move them into a [POM](https://playwright.dev/docs/pom)-style class at the bottom of the file? It's just a coding style we prefer when writing e2e tests to keep them consistent 😆 .

worldomonation replied:
Thanks for suggestion, wasn't aware this is what other tests did. 

Addressed in 071c31b4fe4685041cbd67bfe90e7d5f27131a0c.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+		await editor.insertBlock( { name: 'core/cover' } );
+
+		// Locate the Black color swatch.
+		const blackColorSwatch = page.getByRole( 'button', {
```
Let's query the element from a parent locator (in this case it's the cover block) to make it more readable for regular readers of this code.

worldomonation replied:
Addressed in 6d20287228b3de67acc86cc88e8114e6fe555db2.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+
+		// Get the RGB value of Black.
+		const blackRGB = await blackColorSwatch.evaluate(
+			( node ) => node.style.backgroundColor
```
Instead of relying on the "style" attribute, which is an implementation detail, maybe we should use `window.getComputedStyle` here to get the computed color instead.

worldomonation replied:
Addressed in 6d20287228b3de67acc86cc88e8114e6fe555db2.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+		// The span element implements the colored background.
+		const backgrounDimLocator = page
+			.getByRole( 'document', { name: 'Block: Cover' } )
+			.locator( 'span[aria-hidden="true"]' );
```
Continuing from above, we can directly grab the background color from the cover block if we're using `window.getComputedStyle`.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+
+		// Wait for the img's src attribute to be prefixed with http.
+		// Otherwise, the URL for the img src attribute is that of a placeholder.
+		await coverBlock.locator( `img[src^=http]` ).waitFor();
```
Nit: This works but relies on the implementation detail. What we could do is use `expect().toPass()` to repetitively test a block of assertions until they pass.

worldomonation replied:
Replaced this with a `expect.toPass` wait in 6d20287228b3de67acc86cc88e8114e6fe555db2.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+
+		const [ backgroundDimColor, backgroundDimOpacity ] =
+			await getBackgroundColorAndOpacity(
+				coverBlock.locator( 'span[aria-hidden="true"]' )
```
Hmm, this is a bit awkward, as we're trying to test something that really should be tested by visual testing instead. I guess adding a comment about why we have to use CSS selector above would help the readers?

worldomonation replied:
@kevin940726 Do you think this particular test case is not appropriate in the setting of E2E, because it's testing for specific values of opacity and background color? If so, I'm happy to remove it from the migrated spec.

Side note, computing the style for the Cover Block locator returns an opacity value of 1, while computing the style for the hidden `span` that actually implements the background returns an opacity of 0.5, resulting in a discrepancy.

I am happy to write comments explaining why a CSS span has to be used instead of the cover block. How should I proceed?

kevin940726 replied:
I think adding a comment above will be better for now. In case this was testing for an actual regression and removing it prematurely put us at risk of introducing the bug again. 👍

kevin940726 replied:
Another idea would be to capture the screenshot for this element and do raw pixel manipulation to find the exact color code. This is a lot of work though, and I'm okay to just skip it and do a simpler assertion for this tiny step 😆 . At least until we make this easy to test 😛 .

worldomonation replied:
> do raw pixel manipulation to find the exact color code. This is a lot of work though,

I agree that this is too much work for a test step, especially when considering that checking for a hidden span would be much simpler.

> I think adding a comment above will be better for now

I've added comments in both places where the `span[aria-hidden="true"]` selector is used, explaining the reason.

Fixed in d6a4f4499555c768f60a9ca86ae70cc0746e1332.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+
+		// Activate the paragraph block inside the Cover block.
+		// The name of the block differs depending on whether text has been entered or not.
+		const coverParagraphLocator = page.getByRole( 'document', {
```
If it's inside the cover block, could we use the cover block as the parent locator?

worldomonation replied:
Fixed all instances of this issue in 6d20287228b3de67acc86cc88e8114e6fe555db2.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+		// Establish the existing bounding boxes for the Cover block.
+		const coverBlockBox = await coverBlock.boundingBox();
+
+		expect( coverBlockBox.height ).toEqual( 450 );
```
Is it supposed to be 450? I think this should be an implementation detail?

worldomonation replied:
You're right, I've removed this assertion as it is an implementation detail.

I've instead replaced it with a check that the `coverBlockBox.height` attribute is truthy.

worldomonation replied:
Addressed in 6d20287228b3de67acc86cc88e8114e6fe555db2.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+		await page.mouse.move(
+			coverBlockBox.x,
+			coverBlockBox.y + coverBlockBox.height + 100,
+			{ steps: 5 }
```
Any reason why we have to use `steps` here? I know that some events don't fire expectedly if we're not using steps, is this also the case here?

worldomonation replied:
I ran the spec without the `steps` attribute and it worked just fine. It was a carryover from the Puppeteer spec.

worldomonation replied:
Addressed in 6d20287228b3de67acc86cc88e8114e6fe555db2.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,241 @@
+		await page.mouse.up();
+
+		const newCoverBlockBox = await coverBlock.boundingBox();
+		expect( newCoverBlockBox.height ).toBeGreaterThanOrEqual(
```
Just being curious, why couldn't we just use `toBe` here? Sub-pixel issue?

worldomonation replied:
The boundingbox size that's returned _after_the resize is 562, which is not exactly 100 more than the original size. The original Puppeteer spec tested that the value changed by more than 100 and so I applied a similar line of thinking in migrating the test.

With that said, I went digging to explain the discrepancy.

The first step was to output the bounding box for all involved elements.
Resize handle  { x: 519, y: 683.5, width: 472, height: 23 }
Cover block { x: 519, y: 245.21875, width: 472, height: 450 }

Notice that the coverBlock.height + ( resizeHandle.height / 2 ) + 100 = 561.5 which likely rounds up to 562.

Checking the specific pixel value would obviously be testing an implementation detail. 
In similar situations in the past I would have stuck with either the current implementation, or go with the Puppeteer spec approach of checking that the bounding box changed by at least 100px. What do you think?

kevin940726 replied:
Thanks for digging into the details! I appreciate it a lot!

Would an API like [`toBeCloseTo`](https://jestjs.io/docs/expect#tobeclosetonumber-numdigits) help in this case? (I'm not sure if it's available in Playwright yet though!)

worldomonation replied:
The `toBeCloseTo` API is available in Playwright - it is categorized under [GenericAssertions](https://playwright.dev/docs/api/class-genericassertions#generic-assertions-to-be-close-to) and is implemented similarly to Jest (ie. used to compare floating point numbers).

I gave it a try in the code, by doing some (rather arbitrary) math:

```javascript
	expect( newCoverBlockBox.height ).toBeCloseTo(
			coverBlockBox.height + 100 + coverBlockResizeHandleBox.height / 2,
            0
		);
```

However, I could not get it to work:

```
    Error: expect(received).toBeCloseTo(expected, precision)

    Expected: 561.5
    Received: 562

    Expected precision:    0
    Expected difference: < 0.5
    Received difference:   0.5
```

Now, I realized that I may have migrated this spec wrong and there was a reason why the original Puppeteer spec was using the resize handle as the basis for calculations. So I switched to using the Cover block's resize handle as the basis for pixel calculations and voilà, the resulting Cover block's bounding block size was 550 pixels as expected.

Fixed in 8b19caf1f417412be22351bc0e2fed22991c1be3.

worldomonation committed:
- add comment explaing use of CSS selector for '.components-resizable-box__handle-bottom'.
- move image upload and filename methods previously handled by `getTestImage` into a POM-like function at bottom of file.

worldomonation committed:
- remove testing `test.only`

worldomonation committed:
- define the coverBlock locator after inserting the block, and use that locator as starting point for all other interactions.
- use `getComputedStyle` instad of `node.style.<attribute>`.
- replace implementation detail check on image upload with a `expect.toPass` check.
- remove steps when resizing box.
- replace implementation detail check for coverbox height with truthy check.

worldomonation commented:
@kevin940726 I left two questions - [1](https://github.com/WordPress/gutenberg/pull/45784#discussion_r1189466256) and [2](https://github.com/WordPress/gutenberg/pull/45784#discussion_r1189403023) that I needed further clarification on how to proceed. 

I'm more than happy for more rounds of code reviews should you deem it necessary, so please feel free to take apart the revised spec. Thanks!

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,267 @@
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+async function getBackgroundColorAndOpacity( locator ) {
```
Nit: Let's move this function to the POM at the end too to keep it consistent with others. Think of it as a code convention or coding style 😅

worldomonation replied:
Moved into the `ImageBlockUtils` POM in 7897f28b7eaa3146eb2a023d536bc98c437eb884.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,267 @@
+import path from 'path';
+const fs = require( 'fs/promises' );
+import os from 'os';
+import { v4 as uuid } from 'uuid';
```
Let's keep it consistent and use `require`-style imports.

worldomonation replied:
Sorry, this was my complete oversight in copy-paste. 

One suggestion, given that `require` import is not appropriate, a lint rule may be useful to catch this.

worldomonation replied:
Fixed in d6a4f4499555c768f60a9ca86ae70cc0746e1332.

kevin940726 replied:
I mean use `require` not `import` 😅 . We have to use `require` for now because of some ESM and NodeJS inconsistency 😞 .

worldomonation replied:
> I mean use `require` not `import` 😅 . We have to use `require` for now because of some ESM and NodeJS inconsistency 😞 .

Interesting. The spec seems to run fine in both local and CI environments, but I'll change it over to `require` in the next revision.

worldomonation replied:
Changed to `require` in 7897f28b7eaa3146eb2a023d536bc98c437eb884.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,267 @@
+			.locator( 'img' )
+			.getAttribute( 'src' );
+
+		expect( backgroundImageURL ).toContain( fileBasename );
```
Oops! I mean to just repetitively check if the image url contains the file base name instead. So that we don't have to explicitly wait for an implementation detail like the placeholder url.

worldomonation replied:
Changed to assert and check on the URL scheme in aadc65db9d5009e4b86d011baa31e6fcb95fda26.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,267 @@
+			.click();
+
+		// Select the Cover block from the docuemnt overview.
+		await page.getByRole( 'link', { name: 'Cover' } ).click();
```
We can target the list view using `getByRole('region', { name: 'Document Overview' }` or `getByRole('treegrid', { name: 'Block navigation structure' })`. So it's a bit clearer which link we are clicking on :).

worldomonation replied:
Got it. I have changed all the "raw" selectors to be anchored off a parent element. 

When there are multiple parent elements (like in this case) I chose the one that I think makes sense.

Addressed in b6e5c294eb63aa6f79bc364254f622f75aac533d.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,267 @@
+		await page.getByRole( 'link', { name: 'Cover' } ).click();
+
+		// Ensure there the default value for the minimum height of cover is undefined.
+		await page.getByRole( 'tab', { name: 'Styles' } ).click();
```
Same here and for other selectors below. Let's scope it under a parent locator, which is often a region. A general rule of thumb is that it should be clear by reading the code to know which element we're interacting with. A tab named "Styles" isn't clear enough and there could be multiple tabs with the same name on the same page. Scoping them under a parent helps the reader to locate the element in their head even without interacting with the page themselves.

You can use the devtools to find the accessible parent for an element by using the [accessibility tree view](https://developer.chrome.com/blog/full-accessibility-tree/) (if you're using chrome devtools). For instance, here we know the element is under the region "Editor settings":

![image](https://github.com/WordPress/gutenberg/assets/7753001/6c47e476-ff40-4085-858b-c84be901294e)

worldomonation replied:
Addressed in b6e5c294eb63aa6f79bc364254f622f75aac533d as noted in the comment above this one.

kevin940726 reviewed (commented):
Thanks again for the updates! I left some suggestions and feedback below. Overall this is looking very good though! 💯

worldomonation committed:
- fix the import to use `import`.
- add comment explaining the use of `span[aria-hidden="true"]`.

worldomonation committed:
- anchor selectors on a parent selector.

worldomonation committed:
- use the Cover block's resize handle as basis for calculation when resizing the block.

worldomonation committed:
- remove testing "only"

worldomonation committed:
- check the src repeatedly to contain the fileBasename instead.

worldomonation committed:
- move `getBackgroundColorAndOpacity` into the ImageBlockUtils POM.

kevin940726 commented on test/e2e/specs/editor/blocks/cover.spec.js:
```diff
@@ -0,0 +1,284 @@
+	} );
+} );
+
+class ImageBlockUtils {
```
Maybe we should name it `CoverBlockUtils`? 😆

worldomonation replied:
Makes sense. I thought maybe `ImageBlockUtils` was always used as the name for the POM class, for some reason.

worldomonation replied:
Name changed to `CoverBlockUtils` in 405c70909d0fd3944db7a518805e48256eeb8fd6.

kevin940726 reviewed (approved):
Looking great! Just a small suggestion and I think we're good to go! Thank you so much for the hard work! 🙇

worldomonation committed:
Rename POM to CoverBlockUtils

kevin940726 reviewed (approved):
Thanks again! 💯 

@worldomonation This is really well done! I couldn't have done it better myself! Also congrats on your first merged PR!

worldomonation commented:
> Thanks again! 💯
> 
> @worldomonation This is really well done! I couldn't have done it better myself! Also congrats on your first merged PR!

Thanks @kevin940726 for the many rounds of review. I really enjoyed working on this and would love to pick up a few more in the coming weeks! Thank you again.
